0022-circuit-block.md

```diff
@@ -0,0 +1,355 @@
+
+We do also have a potential option in that core Qiskit can specify a small number of "special case" annotations, which all passes are required to interpret and respect.
+This gets tricky from a backwards compatibility perspective, though - if we add a new one of these special annotations in Qiskit 1.3, say, it might invalidate code that worked happily in Qiskit 1.2.
+
+
+### Serialisation and deserialisation of annotations
+
+> ![NOTE]
+>
+> This RFC doesn't propose a final interface for QPY, just describes the principles of how this will be done.
+> The rest of that design can be a follow-up discussion in Qiskit core.
+> There is a rough preference from the RFC authors to use a model based on the `kwargs` extensions.
+
+If we allow arbitrary annotations, we reasonably need some way of constructing / deconstructing these, without sending arbitrary code to be executed over the wire - that would be a huge security hole.
+This is the problem that QPY solves for circuits.
+The general principle is that we can rely on reconstruction on the far side of a network boundary, if the receiver only runs code that it chose to install ahead of time.
+`pickle` is not suitable for this.
+
+In this model, QPY would need to be extended in a way that allows the serialisation and deserialisation of these attributes.
+Two ways of making this extension would be:
+
+* have a plugin interface to QPY, where installed packages can register additional serialisers / deserialisers for annotations.
+* add additional keyword arguments to the `dump`/`load` functions that allow a user to pass in the additional serialisers / deserialisers manually (like how is done for JSON).
+
+Major risks for this in QPY are that both sides of the deserialiser / serialiser would need to have compatible versions of the custom packages available, and that these custom serialisers would need to uphold the same highly stringent backwards compatibility guarantees that QPY supports.
+Because of these risks, the keyword-argument form might be more appropriate, so it is very clear at the point of call that QPY is being extended, and that the caller must be certain that they trust the custom serialisation/deserialisation.
+
+The same sort of logic can be used for import/export from OpenQASM 3, but since that's lossy and we certainly can never make the same guarantees about backwards compatibility with that, it's rather less of a security and compatibility concern.
+For clarity, OpenQASM 3 already includes the concept of backend-specific annotations, like
+```openqasm3
+OPENQASM 3.0;
+include "stdgates.inc";
+
+@ibm twirl
+box {
+    cx $0, $1;
+}
+```
+
+Certain hardware providers may also have direct-access APIs that do not require submission of the job via Qiskit.
+Custom serialisers/deserialisers of annotations should be aware of this as well; it is effectively the same problem as serialisation to and from OpenQASM 3.
```

Comment:
What if, like OpenQASM, annotations were just text (i.e. `str` type)?

Reply:
In practice, passes will almost certainly want programmatic data structures in their annotations - the "case study" that the primitives team at the bottom gives one or two examples.  If we say "annotations are just text", then every pass that needs to examine the annotation (which could well be _every_ pass, even if just to know it can ignore it) needs to know how to parse at least some of the text.  The aim here is to make it not a problem until we hit the only boundaries where the data structures matter, and to move the sources of truth for encoding / decoding to the places that define them.

Reply:
Ok. I was just wondering if the concept of vendor prefixes could be used here.

Reply:
In practice, I'm sure that the OQ3 part of the custom serialisation/deserialisation absolutely will use vendor prefixes, yeah.